### PR TITLE
asn1.h: Avoid conflict with Windows SDK rpcndr.h

### LINKED ISSUE #14312: Rename ASN1_BOOLEAN variable name in struct asn1_type_st in asn1.h
```
typedef struct asn1_type_st {
    int type;
    union {
        char *ptr;
        ASN1_BOOLEAN boolean;
        ASN1_STRING *asn1_string;
        ASN1_OBJECT *object;
        ASN1_INTEGER *integer;
        ASN1_ENUMERATED *enumerated;
        ASN1_BIT_STRING *bit_string;
        ASN1_OCTET_STRING *octet_string;
        ASN1_PRINTABLESTRING *printablestring;
        ASN1_T61STRING *t61string;
        ASN1_IA5STRING *ia5string;
        ASN1_GENERALSTRING *generalstring;
        ASN1_BMPSTRING *bmpstring;
        ASN1_UNIVERSALSTRING *universalstring;
        ASN1_UTCTIME *utctime;
        ASN1_GENERALIZEDTIME *generalizedtime;
        ASN1_VISIBLESTRING *visiblestring;
        ASN1_UTF8STRING *utf8string;
        /*
         * set and sequence are left complete and still contain the set or
         * sequence bytes
         */
        ASN1_STRING *set;
        ASN1_STRING *sequence;
        ASN1_VALUE *asn1_value;
    } value;
} ASN1_TYPE;
```

**ASN1_BOOLEAN boolean** will create conflicts in windows environment, because there are other places in the Windows SDK that define the **boolean** type. Because of that, the compiler will interpret the boolean variable as a type and two types can't be next to each other. It's a beginner mistake from Windows to create unconditional define, especially a boolean. Compilation errors will be a problem for projects that use OpenSSL on Windows.

Can you please rename it, if it's possible?

### PR DESCRIPTION
**Pull request by t8m**
Fixes #14312

### CODE REVIEW

**Comment by t8m** [comment]
Whether this is really a bug fix and should be applied to all stable branches is debatable.

File: include/openssl/asn1.h.in [hunk 1]
```diff
@@ -472,6 +472,10 @@ typedef const ASN1_ITEM *ASN1_ITEM_EXP (void);
                                 ASN1_STRFLGS_DUMP_UNKNOWN | \
                                 ASN1_STRFLGS_DUMP_DER)
 
+/* rpcndr.h which has a boolean typedef includes this define */
+# ifdef small
+#  define boolean asn1_boolean
+# endif
```
[the remark this answers is not in the document]
Comment (t8m) [reply include/openssl/asn1.h.in:483]: OK, improved.

I do not think the undo test should check for boolean being defined, IMO this is safer.

**Review by tom-cosgrove-arm — APPROVED** [review]

**Comment by nhorman** [comment]
If we're changing the name of the ASN1_BOOLEAN for windows conditional on the inclusion of a specific windows header file, I'm assuming we're ok with the potential ABI breakage in that case.  If so, why not just change it unilaterally so we can avoid the ifdeffery?

**Comment by davidben** [comment]
See https://github.com/openssl/openssl/issues/14312#issuecomment-2052125061. Making an API break here seems to be a bit premature.

**Comment by nhorman** [comment]
FWIW, I don't have the windows sdk in front of me, but the reactos clone of the sdk here:
https://doxygen.reactos.org/dd/d91/rpcndr_8h.html
does show:
```
typedef unsigned char boolean
```
Which definitely seems like its going to hit some namespace conflicts.  Perhaps rpcndr isn't often used?

**Comment by davidben** [comment]
Yeah, the real SDK seems to also use a typedef rather than a macro. I mean, that typedef definitely seems kinda ill-advised, but per that linked comment, it doesn't seem like a typedef actually breaks things. C and C++ seem to be fine with typedefs and field names overlapping. So there's something funny going on with the report and we should get more info from the reporter.

**Comment by t8m** [comment]
Closing this as this shouldn't be needed.

**Pull request closed without merging**